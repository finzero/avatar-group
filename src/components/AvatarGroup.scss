$xsSize: 50px;
$smSize: 100px;
$mdSize: 150px;
$lgSize: 200px; 

.avatarContainer{
  display: flex;
}

.avatarImage{
  border-radius: 50%;
  overflow: hidden;
  border: 5px solid white;
  background: #fff;
}

.avatarImage:not(:first-child){
  margin-left: -50px;
}

.avatarImage img{
  height: inherit;
}

.avatarText{
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background: #eeeeee;
}

/*sizing*/
.xs{
  .avatarImage {
    width: $xsSize;
    height: $xsSize;
  }
  .avatarImage:not(:first-child){
    margin-left: calc($xsSize / 4 * -1);
  }
  .avatarText{
    font-size: 25px;
  }
}

.sm{
  .avatarImage {
    width: $smSize;
    height: $smSize;
  }
  .avatarImage:not(:first-child){
    margin-left: calc($smSize / 4 * -1);
  }
  .avatarText{
    font-size: 45px;
  }
}

.md{
  .avatarImage {
    width: $mdSize;
    height: $mdSize;
  }
  .avatarImage:not(:first-child){
    margin-left: calc($mdSize / 4 * -1);
  }
  .avatarText{
    font-size: 65px;
  }
}

.lg{
  .avatarImage {
    width: $lgSize;
    height: $lgSize;
  }
  .avatarImage:not(:first-child){
    margin-left: calc($lgSize / 4 * -1);
  }
  .avatarText{
    font-size: 75px;
  }
}