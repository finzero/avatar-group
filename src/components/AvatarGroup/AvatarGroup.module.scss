$xsSize: 50px;
$smSize: 100px;
$mdSize: 150px;
$lgSize: 200px;

.avatarContainer {
  display: flex;
}

.avatarImage {
  border-radius: 50%;
  overflow: hidden;
  border: 5px solid white;
  background: #fff;
}

.avatarImage img {
  height: inherit;
}

.avatarText,
.overlimitCount {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background: #e8eaf5;
  color: #787b8a;
}

/*sizing*/
.xs {
  .avatarImage {
    width: $xsSize;
    height: $xsSize;
  }
  .avatarImage:not(:first-child) {
    margin-left: calc($xsSize / 3 * -1);
  }
  .avatarText,
  .overlimitCount {
    font-size: 25px;
  }
}

.sm {
  .avatarImage {
    width: $smSize;
    height: $smSize;
  }
  .avatarImage:not(:first-child) {
    margin-left: calc($smSize / 3 * -1);
  }
  .avatarText,
  .overlimitCount {
    font-size: 40px;
  }
}

.md {
  .avatarImage {
    width: $mdSize;
    height: $mdSize;
  }
  .avatarImage:not(:first-child) {
    margin-left: calc($mdSize / 3 * -1);
  }
  .avatarText,
  .overlimitCount {
    font-size: 50px;
  }
}

.lg {
  .avatarImage {
    width: $lgSize;
    height: $lgSize;
  }
  .avatarImage:not(:first-child) {
    margin-left: calc($lgSize / 3 * -1);
  }
  .avatarText,
  .overlimitCount {
    font-size: 60px;
  }
}
