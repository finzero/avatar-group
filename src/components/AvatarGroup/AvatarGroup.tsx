import style from './AvatarGroup.module.scss';

export type Size = 'xs' | 'sm' | 'md' | 'lg';

export interface IUser {
  image: string;
  name: string;
}

interface IAvatarGroup {
  maxLength: number;
  size: Size;
  users: IUser[];
}

const getInitial = (name: string) => {
  return name
    .split(' ')
    .map((n) => n[0].toUpperCase())
    .join('')
    .substring(0, 2);
};

export default function AvatarGroup({ users, maxLength, size }: IAvatarGroup) {
  return (
    <div
      data-testid="avatar-container"
      className={`${style.avatarContainer} ${style[size]}`}
    >
      {users.slice(0, maxLength).map((user: IUser, idx: number) => (
        <div className={style.avatarImage} key={idx}>
          {user.image ? (
            <img src={user.image} alt={user.name} />
          ) : (
            <div className={style.avatarText}>{getInitial(user.name)}</div>
          )}
        </div>
      ))}
      {users.length - maxLength > 0 && (
        <div className={style.avatarImage}>
          <div className={style.overlimitCount}>
            +{users.length - maxLength}
          </div>
        </div>
      )}
    </div>
  );
}
