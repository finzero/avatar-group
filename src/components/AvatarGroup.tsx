import React, { useEffect, useState } from 'react';
import './AvatarGroup.scss';

export type Size = 'xs' | 'sm' | 'md' | 'lg';

export interface UserProp {
  image: string;
  name: string;
}

interface AvatarGroupProp {
  maxLength: number;
  size: Size;
  users: UserProp[];
}

export default function AvatarGroup({
  users,
  maxLength,
  size,
}: AvatarGroupProp) {
  const [overLimit, setOverLimit] = useState(users.length - maxLength);

  const OverLimit = ({ count }: { count: number }) => {
    return count > 0 ? (
      <div className="avatarImage">
        <div className="overlimitCount">+{count}</div>
      </div>
    ) : null;
  };

  const getInitial = (name: string) => {
    return name
      .split(' ')
      .map((n) => n[0].toUpperCase())
      .join('')
      .substring(0, 2);
  };

  useEffect(() => {
    if (users.length > maxLength) {
      // set over limit number for +{number}
      setOverLimit(users.length - maxLength);
    }
  }, [users, maxLength]);

  return (
    <div data-testid="avatar-container" className={'avatarContainer ' + size}>
      {users.slice(0, maxLength).map((user: UserProp, idx: number) => (
        <div className="avatarImage" key={idx}>
          {user.image ? (
            <img src={user.image} alt={user.name} />
          ) : (
            <div className="avatarText">{getInitial(user.name)}</div>
          )}
        </div>
      ))}
      <OverLimit count={overLimit} />
    </div>
  );
}
