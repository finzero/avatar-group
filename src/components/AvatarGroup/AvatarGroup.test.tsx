import { render, within } from '@testing-library/react';
import '@testing-library/jest-dom';
import AvatarGroup from './AvatarGroup';
import renderer from 'react-test-renderer';

const userData = [
  { name: 'Garnet till Alexandros', image: '' },
  { name: 'edison alexander thomas', image: './cat-6.jpg' },
  { name: 'thomas alfa', image: './cat-7.jpg' },
  { name: 'zidane tribal', image: './cat-5.jpg' },
];

describe(AvatarGroup, () => {
  it('should have class md if size = md', () => {
    const maxLength = 2;
    const { getByTestId } = render(
      <AvatarGroup users={userData} size="md" maxLength={maxLength} />
    );
    const container = getByTestId('avatar-container');
    expect(container.classList.contains('md')).toBe(true);
  });

  it('should render 2 avatar and +2 more avatar if total user is 4 and maxLength is 2', () => {
    const maxLength = 2;
    const { getByTestId } = render(
      <AvatarGroup users={userData} size="lg" maxLength={maxLength} />
    );

    const container = getByTestId('avatar-container');
    const totalImageRendered =
      container.querySelectorAll('.avatarImage').length;
    expect(totalImageRendered).toEqual(3);
  });

  it('should render name initial if there is no image', () => {
    const maxLength = 3;
    const { getByTestId } = render(
      <AvatarGroup users={userData} size="md" maxLength={maxLength} />
    );
    const container = getByTestId('avatar-container');
    const userInitial = 'GT'; //  { name: 'Garnet till Alexandros', image: '' }
    expect(within(container).queryByText(userInitial)).toBeTruthy();
  });

  it('should render all avatar if total avatar <= maxLength', () => {
    const maxLength = 4;
    const { getByTestId } = render(
      <AvatarGroup users={userData} size="md" maxLength={maxLength} />
    );
    const container = getByTestId('avatar-container');
    const totalImageRendered =
      container.querySelectorAll('.avatarImage').length;
    expect(totalImageRendered).toEqual(4);
  });

  it('should contain + mark if total avatar > maxLength', () => {
    const maxLength = 2;
    const { getByTestId } = render(
      <AvatarGroup users={userData} size="md" maxLength={maxLength} />
    );
    const container = getByTestId('avatar-container');
    const overlimitCount = container.querySelector(
      '.avatarImage .overlimitCount'
    )?.textContent;
    expect(overlimitCount).toContain('+');
  });

  it('should render correctly', () => {
    const maxLength = 2;
    const tree = renderer
      .create(<AvatarGroup users={userData} size="sm" maxLength={maxLength} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
