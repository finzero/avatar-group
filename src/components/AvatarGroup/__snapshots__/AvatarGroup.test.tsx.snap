// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup should render correctly 1`] = `
<div
  className="avatarContainer sm"
  data-testid="avatar-container"
>
  <div
    className="avatarImage"
  >
    <div
      className="avatarText"
    >
      GT
    </div>
  </div>
  <div
    className="avatarImage"
  >
    <img
      alt="edison alexander thomas"
      src="./cat-6.jpg"
    />
  </div>
  <div
    className="avatarImage"
  >
    <div
      className="overlimitCount"
    >
      +
      2
    </div>
  </div>
</div>
`;
